trade-finance-manager-ui/server/controllers/case/parties/index.js

```diff
@@ -33,11 +36,17 @@ const getAllParties = async (req, res) => {
 
     const canEdit = userCanEdit(user);
 
+    const { submissionType } = deal.dealSnapshot;
+
+    const successMessage =
```

Comment:
Can we use middleware to do this instead of repeating the same thing in 7 controllers?

Middleware can check if the current route is in array of relevant routes, if so, can do e.g

```ts
const successBannerMessage = await getSuccessBannerMessage(submissionType, user, userToken, dealId, getUkefDealId(deal.dealSnapshot));

const flashSuccessMessage = getFlashSuccessMessage(req);

req.locals.successMessage = successBannerMessage || flashSuccessMessage;

return next();

Reply:
The `getSuccessBannerMessage` parameters come from the deal object & it makes a request to the cancellation endpoint to determine if the deal is scheduled for cancellation or not.

We could add lots more middleware to commonise this (as has been discussed before) but it would be a large change of pattern & definitely out of scope for this ticket

Reply:
I am concerned that we are otherwise introducing tech debt by repeating the same thing 7x times.

Appreciate we've had discussions about this before - would be great to move e.g `await api.getDeal` into middleware.

I propose that as an interim step we introduce some middleware purely for the success banner - for these routes, that calls `api.getDeal` and handles the success banner - then we can eventually create some `getDeal` middleware to do this for all routes. I am thinking something like this:

```ts
const RELEVANT_ROUTES = [
  '/summary',
  '/x-route/activity',
  '/etc',
];

const shouldExecute = (url: string) => {
  if (url.includes('/case') && RELEVANT_ROUTES.includes(url)) {
    return true;
  }

  return false;
};

const cancellationSuccessBannerMiddleware = async (req, res, next) => {
  if (shouldExecute(req.originalUrl)) {
    const { user, userToken } = req.session;

    const { dealId, dealSnapshot, submissionType } = await api.getDeal(req.params._id, userToken);

    const successBannerMessage = await getSuccessBannerMessage(submissionType, user, userToken, dealId, getUkefDealId(dealSnapshot));

    res.locals.successMessage = successBannerMessage || flashSuccessMessage;

    return next();
  }

  return next();
};
```

Then:

1. We don't need `api.getDeal` calls in these controllers.
2. We don't have the same logic repeated 7 times in each controller and unit test.
3. If we need to update the logic, we only need to do it in one place.
4. We don't introduce tech debt.
  - I know we'd eventually need to remove `api.getDeal` from the proposed interim middleware, but this is miniscule.
6. For each relevant controller (that's currently updated in this PR), they can have `api.getDeal` removed and simply consume `res.locals.successMessage`.

I think it is cleaner and more scalable to do this as part of this PR for the reasons above. What do you think?

Reply:
This feels like a large change & I've explicitly been told not to introduce new patterns. I'd want to have a chat with @abhi-markan  & the rest of the devs before making db calls in middleware.

I think in the meantime the techdebt incurred by having the same 2 function calls repeated in each controller isnt too large

Reply:
Thanks all.

@AlexBramhill I don't entirely agree RE middleware. I've worked on plenty of applications that have business logic in middleware. A simple example would be user/auth checks - this is business logic and should absolutely be in middleware.

I know this could seem "magical" as you describe, or confusing, for new devs, but having the same logic repeated in various places, instead of having 1x DRY instance, can be equally confusing and also unscalable - it is a bad practice.

A new helper like you suggested seems great and much preferred to the current implementation.

I would like us to explore and implement a middleware to get a deal though and other common api fetches; One for another day.

Reply:
Completely agree re middleware -- I probably oversimplified too much. My general thoughts align with [nexts](https://nextjs.org/docs/pages/building-your-application/routing/middleware).

If we are making the pattern that middleware sources all data we need for an endpoint that's fine (and feels okay tbh) -- but I think we need to clearly define scope as not to end up with everything in middleware, and figure out how we're going to do nice types for it.

Reply:
Agreed @AlexBramhill I think only "general" or "common" things like getting a deal or getting a facility should be in middleware and locked down but a list of specific routes and methods. For example a lot of GET routes will have similar POST routes for form validation, some POSTs might only need deal/facility data if there are validation errors, so that should be handled on the controller level

One for another day 😄